src/FederationSetup/RecoveryTransactionCreator.cs

```diff
@@ -0,0 +1,206 @@
+
+namespace FederationSetup
+{
+    public class FundsRecoveryTransactionModel
+    {
+        public Network Network { get; set; }
+
+        public bool IsSideChain { get; set; }
+
+        public PubKey PubKey { get; set; }
+
+        public Script RedeemScript { get; set; }
+
+        public BitcoinAddress oldMultisigAddress { get; set; }
+
+        public BitcoinAddress newMultisigAddress { get; set; }
+
+        public Transaction tx { get; set; }
+
+        public string ChainType => (this.IsSideChain ? "Sidechain" : "Mainchain");
+
+        public void DisplayInfo()
+        {
+            Console.WriteLine($"Old {this.Network} P2SH: " + this.oldMultisigAddress.ScriptPubKey);
+            Console.WriteLine($"Old {this.Network} multisig address: " + this.oldMultisigAddress);
+
+            Console.WriteLine($"New {this.Network} P2SH: " + this.newMultisigAddress.ScriptPubKey);
+            Console.WriteLine($"New {this.Network} multisig address: " + this.newMultisigAddress);
+
+            Console.WriteLine($"The transaction has been written to the data directory.");
+            Console.WriteLine($"Amount of moving funds: {this.tx.Outputs.Sum(o => o.Value.ToDecimal(MoneyUnit.BTC))}.");
+        }
+
+        public string Signature()
+        {
+            PayToMultiSigTemplateParameters multisigParams = PayToMultiSigTemplate.Instance.ExtractScriptPubKeyParameters(this.RedeemScript);
+
+            int index = multisigParams.PubKeys.IndexOf(this.PubKey);
+
+            TransactionSignature signature = PayToMultiSigTemplate.Instance.ExtractScriptSigParameters(this.Network, new Script(tx.Inputs[0].ScriptSig.ToOps().SkipLast(1)))[index];
+
+            return signature.ToString();
+        }
+    }
+
+    public class RecoveryTransactionCreator
+    {
+        /// <summary>
+        /// Creates a transaction to transfers funds from an old federation to a new federation.
+        /// </summary>
+        /// <param name="isSideChain">Indicates whether the <paramref name="network"/> is the sidechain.</param>
+        /// <param name="network">The network that we are creating the recovery transaction for.</param>
+        /// <param name="counterChainNetwork">The counterchain network.</param>
+        /// <param name="dataDirPath">The root folder containing the old federation.</param>
+        /// <param name="redeemScript">The new redeem script.</param>
+        /// <param name="password">The password required to generate transactions using the federation wallet.</param>
+        /// <param name="txTime">Any deposits beyond this UTC date will be ignored when selecting coin inputs.</param>
+        /// <returns>A funds recovery transaction that moves funds to the new redeem script.</returns>
+        public FundsRecoveryTransactionModel CreateFundsRecoveryTransaction(bool isSideChain, Network network, Network counterChainNetwork, string dataDirPath, Script redeemScript, string password, DateTime txTime)
```

Comment:
Do I understand correctly that this will create the transaction the first time.
Then we need to circulate it to the necessary signers. 
Then deposit their signed responses in the same folder.
And run it again?

Reply:
Correct.